### PR TITLE
Add Microsoft.ContainerInstance API spec 2017-08-01-preview

### PR DESCRIPTION
**Pull request by yolocs**
### Known issues
- No patch operation
- No operations API

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json [hunk 1]
```diff
@@ -0,0 +1,819 @@
...
+        },
+        "ports": {
+          "description": "The ports exposed by this container",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/ContainerPort"
+          }
+        },
+        "environmentVariables": {
+          "type": "array",
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json]: for the record: description missing

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json [hunk 2]
```diff
@@ -0,0 +1,819 @@
...
+          "type": "string",
+          "description": "The state of this container"
+        },
+        "startTime": {
+          "type": "string",
+          "format": "date-time",
+          "description": "Date/time when the container state started."
+        },
+        "exitCode": {
+          "type": "integer"
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json]: for the record: description missing

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json [hunk 3]
```diff
@@ -0,0 +1,819 @@
...
+            },
+            "currentState": {
+              "description": "Current container state.",
+              "$ref": "#/definitions/ContainerState"
+            },
+            "previousState": {
+              "description": "Previous container state.",
+              "$ref": "#/definitions/ContainerState"
+            },
+            "events": {
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json]: for the record: description missing

**Comment by olydis** [comment]
There are ~13 descriptions missing. I heard of the time pressure concerning this PR, but missing documentation makes for an awful UX. If not with this PR, please fix this ASAP.

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json (hunk 3 above)
Comment (yolo3301) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json]: I will add these missing descriptions asap in a followup PR.

**Comment by olydis** [comment]
The two operations returning `ContainerGroupListResult` were not marked as `x-ms-pageable` although they clearly should be. See https://github.com/Azure/azure-rest-api-specs/blob/current/specification/compute/resource-manager/Microsoft.ContainerService/2017-01-31/containerService.json#L61 for an example. (This allows code gen to generate convenience wrappers/methods for dealing with the pages of `ContainerGroupListResult`). Note: I made these changes in #1470 to probe CI and generate code from it, feel free to copy the fixes.

**Comment by seanmck** [comment]
@olydis : ACK on the descriptions. We will fast-follow this PR with another one that takes care of those.

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsListAll.json [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                        "port": 80
+                      }
+                    ],
+                    "resources": null
+                  }
+                }
+              ],
+              "imageRegistryCredentials": null,
+              "ipAddress": {
+                "ip": "10.0.0.0",
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsListAll.json]: you marked `type` as required, but haven't provided it here. Check whether the Swagger is correct or this example and fix either accordingly

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsListAll.json [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                    "protocol": "TCP"
+                  }
+                ]
+              },
+              "osType": "Linux",
+              "provisioningState": "Succeeded",
+              "restartPolicy": null,
+              "state": null
+            },
+            "resourceGroup": "demo",
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsListAll.json]: this property is not documented in the Swagger (so it won't be accessible in statically typed SDKs), just making sure you didn't accidentally miss it

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsList.json [hunk 6]
```diff
@@ -0,0 +1,55 @@
+{
+  "parameters": {
+    "subscriptionId": "subid",
+    "api-version": "2017-08-01-preview"
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsList.json]: `resourceGroupName` parameter missing!

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsGet.json [hunk 7]
```diff
@@ -0,0 +1,62 @@
+{
+  "parameters": {
+    "subscriptionId": "subid",
+    "api-version": "2017-08-01-preview"
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsGet.json]: `resourceGroupName` parameter missing!

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsCreateOrUpdate.json [hunk 8]
```diff
@@ -0,0 +1,148 @@
+{
+  "parameters": {
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsCreateOrUpdate.json]: `containerGroupName` missing

**Review by olydis — COMMENTED** [review]
Issues found:
- documentation missing (deferred)
- `x-ms-pageable` annotations missing (pull from #1470, also adds `tag` for code generation)
- inaccurate examples (please revise if the mismatch indicates an issue in the Swagger. If so, fix with this PR, otherwise, fix the example later, but ASAP)
- `modelAsString: false` in cases that very much look like they should use `true`

**Comment by olydis** [comment]
another important one: `"modelAsString": false` is *highly discouraged* since it makes any changes to the enum a breaking change! I strongly advise changing all 4 occurrences to `true`.

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json [hunk 9]
```diff
@@ -0,0 +1,819 @@
...
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/Port"
+          }
+        },
+        "type": {
+          "type": "string",
+          "description": "Determines whether the IP is exposed to the public internet.",
+          "enum": [
+            "Public"
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json]: are you accidentally missing another option?

Comment (seanmck) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json]: No, public IP is all that's supported currently. Eventually, we will add private IPs on a customer vnet.

**Comment by yolo3301** [comment]
@olydis I think I see more `"modelAsString": false` than `true` in this repo. That's why I set them to false.

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json (hunk 9 above)
Comment (yolo3301) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json]: No. That's the only option now.

**Comment by olydis** [comment]
@yolo3301 We learnt from past mistakes. Changing the existing enums from `false` to `true` would be a breaking change too, so we leave them for now. But for new enums, go with `true`. This is actually an ARM guideline now.

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json (hunk 3 above)
Comment (yolo3301) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/containerInstance.json]: Added all missing descriptions except one for "properties"

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsGet.json (hunk 7 above)
Comment (yolo3301) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsGet.json]: Added

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsListAll.json (hunk 4 above)
Comment (yolo3301) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsListAll.json]: Added

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsListAll.json (hunk 5 above)
Comment (yolo3301) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsListAll.json]: Added

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsList.json (hunk 6 above)
Comment (yolo3301) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsList.json]: Added

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsCreateOrUpdate.json (hunk 8 above)
Comment (yolo3301) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsCreateOrUpdate.json]: Added

**Comment by yolo3301** [comment]
@olydis I should have addressed all your comments. Thanks for the feedback.

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsListAll.json [hunk 10]
```diff
@@ -8,50 +8,71 @@
       "body": {
         "value": [
           {
-            "id": "/subscriptions/subid/resourceGroups/demo/providers/Microsoft.ContainerInstance/containerGroups/demo1",
-            "location": "west us",
-            "name": "demo1",
-            "properties": {
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsListAll.json]: why did you remove this nesting? This was correct (or at least according to the Swagger) before! Are you sure that your API returns *this*???

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsList.json [hunk 11]
```diff
@@ -1,52 +1,79 @@
...
   },
   "responses": {
     "200": {
       "body": {
         "value": [
           {
-            "id": "/subscriptions/subid/resourceGroups/demo/providers/Microsoft.ContainerInstance/containerGroups/demo1",
-            "location": "west us",
-            "name": "demo1",
-            "properties": {
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsList.json]: feels like the nesting was correct, see other comment

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsGet.json [hunk 12]
```diff
@@ -1,61 +1,77 @@
...
+    "resourceGroupName": "demo",
     "api-version": "2017-08-01-preview"
   },
   "responses": {
     "200": {
       "body": {
-        "id": "/subscriptions/subid/resourceGroups/demo/providers/Microsoft.ContainerInstance/containerGroups/demo1",
-        "location": "west us",
-        "name": "demo1",
-        "properties": {
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsGet.json]: feels like the nesting was correct, see other comment

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsCreateOrUpdate.json [hunk 13]
```diff
@@ -48,100 +50,96 @@
...
+        "type": "Microsoft.ContainerInstance/containerGroups",
+        "volumes": null
       }
     },
     "201": {
       "body": {
-        "id": "/subscriptions/subid/resourceGroups/demo/providers/Microsoft.ContainerInstance/containerGroups/demo1",
-        "location": "west us",
-        "name": "demo1",
-        "properties": {
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsCreateOrUpdate.json]: feels like the nesting was correct, see other comment

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsCreateOrUpdate.json [hunk 14]
```diff
@@ -48,100 +50,96 @@
   "responses": {
     "200": {
       "body": {
-        "id": "/subscriptions/subid/resourceGroups/demo/providers/Microsoft.ContainerInstance/containerGroups/demo1",
-        "location": "west us",
-        "name": "demo1",
-        "properties": {
```
Comment (olydis) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsCreateOrUpdate.json]: feels like the nesting was correct, see other comment

**Comment by olydis** [comment]
@yolo3301 Most of your changes look good, however you have also changed some examples in a way that they no longer match the Swagger! Was this intentional? Which one is correct now, examples or Swagger? Please clarify and adjust.

File: specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsListAll.json (hunk 10 above)
Comment (yolo3301) [inline specification/compute/resource-manager/Microsoft.ContainerInstance/2017-08-01-preview/examples/ContainerGroupsListAll.json]: This is the output when I test in with CLI. I think it's because I added `"x-ms-client-flatten": true`

**Comment by yolo3301** [comment]
@olydis I added flag `"x-ms-client-flatten": true`. Checked some examples in other folders, I see the nest is also removed. Am I doing it right?

**Comment by olydis** [comment]
> Checked some examples in other folders, I see the nest is also removed.

Oh dear, that's an issue that should be tracked somewhere (our validation tools also go crazy on it, so not sure how that other stuff got in!)

Note that `"x-ms-client-flatten": true` is purely about generated code and *not* about the wire format. The examples also represent sample requests and response as you would encounter them on the wire, so `x-ms-client-flatten` should have absolutely no influence on them.

**Review by olydis — APPROVED** [review]
awaiting fixes to examples, otherwise good to go 👍

**Comment by seanmck** [comment]
Awesome. Thanks for the quick but thorough review on this, @olydis.

**Comment by olydis** [comment]
@seanmck sure thing! Well, as thorough as possible in the time given, but I think it's okay for `preview` 😉

**Comment by yolo3301** [comment]
@olydis Examples updated.
